src/java.base/share/classes/jdk/internal/foreign/abi/ppc64/aix/AixPPC64VaList.java

```diff
@@ -33,123 +33,247 @@
-    }
-
-    static AixPPC64VaList.Builder builder(MemorySession session) {
-        return new AixPPC64VaList.Builder(session);
-    }
-
-    public static non-sealed class Builder implements VaList.Builder {
-
-        public Builder(MemorySession session) {
-            throw new InternalError("Builder() is not yet implemented"); //$NON-NLS-1$
-        }
-
-        @Override
-        public Builder addVarg(ValueLayout.OfInt layout, int value) {
-            throw new InternalError("addVarg() is not yet implemented"); //$NON-NLS-1$
-        }
-
-        @Override
-        public Builder addVarg(ValueLayout.OfLong layout, long value) {
-            throw new InternalError("addVarg() is not yet implemented"); //$NON-NLS-1$
-        }
-
-        @Override
-        public Builder addVarg(ValueLayout.OfDouble layout, double value) {
-            throw new InternalError("addVarg() is not yet implemented"); //$NON-NLS-1$
-        }
-
-        @Override
-        public Builder addVarg(ValueLayout.OfAddress layout, Addressable value) {
-            throw new InternalError("addVarg() is not yet implemented"); //$NON-NLS-1$
-        }
-
-        @Override
-        public Builder addVarg(GroupLayout layout, MemorySegment value) {
-            throw new InternalError("addVarg() is not yet implemented"); //$NON-NLS-1$
-        }
-
-        public VaList build() {
-            throw new InternalError("build() is not yet implemented"); //$NON-NLS-1$
-        }
-    }
+	public static final Class<?> CARRIER = MemoryAddress.class;
+
+	/* Every primitive/pointer occupies 8 bytes and structs are aligned
+	 * with 8 bytes in the total size when stacking the va_list buffer.
+	 */
+	private static final long VA_LIST_SLOT_BYTES = 8;
+	private static final VaList EMPTY = new SharedUtils.EmptyVaList(MemoryAddress.NULL);
+
+	private MemorySegment segment;
+	private final MemorySession session;
+
+	private AixPPC64VaList(MemorySegment segment, MemorySession session) {
+		this.segment = segment;
+		this.session = session;
+	}
+
+	public static final VaList empty() {
+		return EMPTY;
+	}
+
+	@Override
+	public int nextVarg(ValueLayout.OfInt layout) {
+		return Math.toIntExact((long)readArg(layout));
+	}
+
+	@Override
+	public long nextVarg(ValueLayout.OfLong layout) {
+		return (long)readArg(layout);
+	}
+
+	@Override
+	public double nextVarg(ValueLayout.OfDouble layout) {
+		return (double)readArg(layout);
+	}
+
+	@Override
+	public MemoryAddress nextVarg(ValueLayout.OfAddress layout) {
+		return (MemoryAddress)readArg(layout);
+	}
+
+	@Override
+	public MemorySegment nextVarg(GroupLayout layout, SegmentAllocator allocator) {
+		Objects.requireNonNull(allocator);
+		return (MemorySegment)readArg(layout, allocator);
+	}
+
+	private Object readArg(MemoryLayout argLayout) {
+		return readArg(argLayout, SharedUtils.THROWING_ALLOCATOR);
+	}
+
+	private Object readArg(MemoryLayout argLayout, SegmentAllocator allocator) {
+		Objects.requireNonNull(argLayout);
```

Comment:
I see `Objects.requireNonNull()` is checking `allocator` at line 104, but it is checking `argLayout` only here.

Reply:
That's correct as we only check the passed-in allocator in the case `struct` given `SharedUtils.THROWING_ALLOCATOR` is set as the default allocator for primitives/pointer in `readArg(MemoryLayout argLayout)`.

I've updated the code here to put these checks together to avoid any confusion.